[WIP] Simple example with multi NFT markers, to be improved

kalwalt opened the pull request:
With this PR we will add a simple example with the new Multi NFT feature, and update all the other examples to the new methods (if it is needed).
From discussion in issue #8

List of changes:
- added **nft_multi.html** example in **nft_improved_worker** folder
- created `artoolkit.multi_worker.js` and `multi_nft_worker.js`
- updated **nft_improved_worker/index.html** with the new example
- two new NFT marker kuva and chalk_multi in DataNFT folder
- cleaned the code: deleted references to container and greyCover (where not needed anymore)

kalwalt commented:
All other NFT examples works without any modification, tested all the examples. :smile:

kalwalt commented:
I used for multi nft example three images inside the Data folder: pinball.jpg, chalk_multi.jpg and kuva.jpg. This without uploading other images and enlaring the repository. They have all 5 stars (Confidence), pretty good for tracking. Used simple 3D meshes (sphere, cube, cone) to keep the code very basic. Of course you can display onto the marker any object you imagine. 
I cleaned also a bit the nft examples code, there was old references to `container` and `greyCover` that are not necessary.
If you find something that need to be changed, let me know. I think that PR is almost ready to be merged.

hiukim commented on js/artoolkit.multi_worker.js:
```diff
@@ -0,0 +1,61 @@
+            markerResult = {type: "found", index: JSON.stringify(ev.data.index),  matrixGL_RH: JSON.stringify(ev.data.matrixGL_RH), proj: JSON.stringify(cameraMatrix)};
+        });
+
+        ar.loadNFTMarkers(msg.marker, function (markerId) {
```
what's the callback `markerId` here?

kalwalt replied:
given the markerId it tracks the marker, we have already used this pattern.

hiukim replied:
msg.marker seems to be a list. the markerId refers to the whole list rather than an individual marker?

kalwalt replied:
yes it is refering to the array/list of markers

kalwalt replied:
printing markerId, it will give you the whole list

kalwalt replied:
Thinking now, i could ahnge this to track every single marker inside this routine, and esnding a postmessage for every single marker? What is better?

kalwalt replied:
i will do a snippet, as example.

kalwalt replied:
We have trhee NTF markers, so i can track each one? Do you think it is more correct this code?

```javascript
ar.loadNFTMarkers(msg.marker, function (markerId) {
            ar.trackNFTMarkerId(markerId[0], 2);
            ar.trackNFTMarkerId(markerId[1], 2);
            ar.trackNFTMarkerId(markerId[2], 2);
            console.log("loadNFTMarker -> ", markerId);
            postMessage({type: "endLoading", end: true})
        });
```

hiukim replied:
I see. I think what you have originally is fine. You probably want to track all markers anyway. Perhaps, naming them `msg.makers` and `markerIds` can avoid some confusion. But it's just a minor issue. 

I have no problem with this PR. I haven't tried running the example, but the code looks fine with me. 

Thanks for the effort!

kalwalt replied:
@hiukim thanks for the review! i will think about this last considerations, and yes you are right i will rename with a more convenient name.

kalwalt replied:
@hiukim at the end i changed the names... :slightly_smiling_face:

kalwalt commented:
Maybe will be better to use the Wasm version of the library instead of the asm.js version (artoolkit.min.js). I think i will do this.

kalwalt commented:
I added alsot he WASM example, but i will re-check all the code in the next days. Anyway it should be almost ready. :smiley:

kalwalt commented:
I need to fix a last issue on the wasm version, will do in the next days...

kalwalt commented:
https://github.com/webarkit/jsartoolkit5/pull/10/commits/cbd2d06cbbd914dc7083cb553f0b98cc2fb387d9 fixed the issue, finale re-check in the next days.

sunshine8550 commented:
I'm very pleased to see you working on NFT.I tried the example of multi_nft.html,and I found that when I identified the target, the memory rose so fast that it crashed.I don't know if you have that problem?

kalwalt commented:
@sunshine8550 thank you for reporting this issue. No, i didn't had this problem, but i hadn't tested deeply. Can you tell me your specs ( Os and device) ? Thank you.

kalwalt commented:
@sunshine8550 have you tried the asm or WASM version? I found a bug on the WASM version, maybe that caused the issue?

kalwalt commented:
i receive this error in the console, while detecting a NFT marker:
```
VM77:1 Uncaught SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at found (multi_nft_wam_worker.js:185)
    at Worker.worker.onmessage (multi_nft_wam_worker.js:165)
```

kalwalt commented:
It seems like that some commits are lost? I need to fix this.

kalwalt commented:
@sunshine8550  can you try the multi WASM NFT example? I fixed a bug that maybe caused that issue ( my stupid mistake !) 🙂

sunshine8550 commented:
> @sunshine8550 thank you for reporting this issue. No, i didn't had this problem, but i hadn't tested deeply. Can you tell me your specs ( Os and device) ? Thank you.
@kalwalt Thank you for your reply! I use Windows 7+Firefox

sunshine8550 commented:
> @sunshine8550 can you try the multi WASM NFT example? I fixed a bug that maybe caused that issue ( my stupid mistake !)

@kalwalt ok , I'll test that out later

sunshine8550 commented:
> @sunshine8550 can you try the multi WASM NFT example? I fixed a bug that maybe caused that issue ( my stupid mistake !)



> @sunshine8550 can you try the multi WASM NFT example? I fixed a bug that maybe caused that issue ( my stupid mistake !)

@kalwalt This example works fine! Finally, I can use multi NFT in my project. I'm so excited!Thank you for your efforts!

sunshine8550 commented:
> @kalwalt， 我很高兴看到您正在NFT上工作。我尝试了multi_nft.html的示例，我发现当我确定目标时，内存上升得如此之快，以至于崩溃了。我不知道您是否拥有这个问题吗？

@kalwalt  I found the reason.In artoolkit.multi_worker.js lin32 , memory runs out when your console prints 'ev'.

kalwalt commented:
Ok i will look at. I think i can remove It, should be only for testing. 🙂

kalwalt commented:
> > @sunshine8550 can you try the multi WASM NFT example? I fixed a bug that maybe caused that issue ( my stupid mistake !)
> 
> > @sunshine8550 can you try the multi WASM NFT example? I fixed a bug that maybe caused that issue ( my stupid mistake !)
> 
> @kalwalt This example works fine! Finally, I can use multi NFT in my project. I'm so excited!Thank you for your efforts!

You should thank @hiukim, he did the hard work to implement this feature. I only did the examples! 😄 
I removed that log causing the issue, should be fine now.

sunshine8550 commented:
@hiukim Thank you for your hard work.I have been waiting for the multi NFT for a long time, and now my project can proceed smoothly.

kalwalt commented:
I think that now can be merged, i will do the last tests in the next days. 🙂
